Frameworks/UIKit/NSTextContainer.mm

```diff
@@ -75,25 +119,69 @@ - (CGRect)lineFragmentRectForProposedRect:(CGRect)proposed
+            CGRect lineIntersection{ { _exclusionPathBoundingRects[i].origin.x, proposed.origin.y },
+                                     { _exclusionPathBoundingRects[i].size.width, proposed.size.height } };
+
+            // TODO 1394: Convert to CGPathContainsPoint for CGD2D
+            // Find the leftmost and rightmost points of the exclusion zone intersecting the same horizontal region as the proposed rect
+            lineIntersection.origin.x = __GetXPositionIntersectingZone(lineIntersection, _exclusionContexts[i].get(), padding);
+            lineIntersection.size.width =
+                __GetXPositionIntersectingZone({ { lineIntersection.origin.x + lineIntersection.size.width, lineIntersection.origin.y },
+                                                 lineIntersection.size },
+                                               _exclusionContexts[i].get(),
+                                               -padding) -
+                lineIntersection.origin.x;
+            exclusionRectsForHorizontal.emplace_back(lineIntersection);
         }
+    }
 
-        if (remainingRect) {
-            *remainingRect = CGRectMake(ret.origin.x + ret.size.width + self.lineFragmentPadding,
-                                        ret.origin.y,
-                                        _size.width - ret.origin.x - ret.size.width,
-                                        ret.size.height);
-            for (UIBezierPath* path in(NSArray*)_exclusionPaths) {
-                CGRect shapeRect = _CGPathFitRect(path.CGPath, *remainingRect, _size, self.lineFragmentPadding);
-                *remainingRect = CGRectIntersection(shapeRect, *remainingRect);
-            }
+    // Remove any exclusion zones not intersecting horizontal with proposed rect
+    std::remove_if(exclusionRectsForHorizontal.begin(), exclusionRectsForHorizontal.end(), [](CGRect rect) {
+        return rect.size.width <= 0;
+    });
+
+    if (exclusionRectsForHorizontal.size() == 0L) {
+        return ret;
+    }
+
+    // Sort by X origin to prevent skipping possible areas not covered by an exclusion zone
+    std::sort(exclusionRectsForHorizontal.begin(), exclusionRectsForHorizontal.end(), [](CGRect a, CGRect b) {
+        return a.origin.x < b.origin.x;
+    });
+    ret = __FirstPossibleRectForProposed(ret, exclusionRectsForHorizontal, padding, _size.width);
+
+    if (remainingRect) {
+        CGFloat newXPosition = ret.origin.x + ret.size.width + 2 * padding;
+        if (newXPosition < _size.width) {
+            *remainingRect =
+                __FirstPossibleRectForProposed(CGRectMake(newXPosition, ret.origin.y, _size.width - newXPosition, ret.size.height),
+                                               exclusionRectsForHorizontal,
+                                               padding,
+                                               _size.width);
         }
     }
```

Comment:
yes it's two ifs and rather than going with preference, we should look at code readability.
Most of this code is already heavy and it would make sense reduce the clutter. 

---
In reply to: [88341543](https://github.com/Microsoft/WinObjC/pull/1398#discussion_r88341543) [](ancestors = 88341543)